src/script/view_model/content/AppLockViewModel.ts

```diff
@@ -0,0 +1,290 @@
+      () => this.passwordRegex.test(this.setupPasswordB()) && this.setupPasswordB() === this.setupPasswordA(),
+    );
+    this.unlockError = ko.observable('');
+    this.wipeError = ko.observable('');
+    this.appObserver = new MutationObserver(mutationRecords => {
+      const [{attributeName}] = mutationRecords;
+      if (attributeName === 'style') {
+        amplify.publish(WebAppEvents.LIFECYCLE.SIGN_OUT, SIGN_OUT_REASON.USER_REQUESTED);
+      }
+    });
+    this.modalObserver = new MutationObserver(() => {
+      const modalInDOM = document.querySelector('[data-uie-name="applock-modal"]');
+      if (!modalInDOM) {
+        amplify.publish(WebAppEvents.LIFECYCLE.SIGN_OUT, SIGN_OUT_REASON.USER_REQUESTED);
+      }
+    });
+    if (isAppLockEnabled()) {
+      this.showAppLock();
+      if (isUnfocusAppLockEnabled()) {
+        document.addEventListener('visibilitychange', this.handleVisibilityChange, false);
+      }
+      this.startPassphraseObserver();
+    }
+    amplify.subscribe(WebAppEvents.PREFERENCES.CHANGE_APP_LOCK_PASSPHRASE, this.changePassphrase);
+  }
+
+  getStored = () => this.localStorage.getItem(this.storageKey());
+
+  setCode = async (code: string) => {
+    this.stopPassphraseObserver();
+    await sodium.ready;
+    const hashed = sodium.crypto_pwhash_str(
+      code,
+      sodium.crypto_pwhash_OPSLIMIT_INTERACTIVE,
+      sodium.crypto_pwhash_MEMLIMIT_INTERACTIVE,
+    );
+    this.localStorage.setItem(this.storageKey(), hashed);
+    this.startPassphraseObserver();
+  };
+
+  onClosed = () => {
+    this.state(APPLOCK_STATE.NONE);
+    this.setupPasswordA('');
+    this.setupPasswordB('');
+  };
+
+  handlePassphraseStorageEvent = ({key, oldValue}: StorageEvent) => {
+    if (key === this.storageKey()) {
+      this.localStorage.setItem(this.storageKey(), oldValue);
+    }
+  };
+
+  startPassphraseObserver = () => window.addEventListener('storage', this.handlePassphraseStorageEvent);
+  stopPassphraseObserver = () => window.removeEventListener('storage', this.handlePassphraseStorageEvent);
+
+  startObserver = () => {
+    afterRender(() => {
+      this.modalObserver.observe(document.querySelector('#wire-main'), {
+        childList: true,
+        subtree: true,
+      });
+      this.appObserver.observe(document.querySelector('#app'), {attributes: true});
+    });
+  };
+
+  stopObserver = () => {
+    this.modalObserver.disconnect();
+    this.appObserver.disconnect();
+  };
+
+  handleVisibilityChange = () => {
+    window.clearTimeout(this.unfocusTimeOutId);
+    const isHidden = document.visibilityState === 'hidden';
+    if (isHidden) {
+      this.unfocusTimeOutId = window.setTimeout(this.showAppLock, getUnfocusAppLockTimeOut() * 1000);
+    }
+  };
+
+  startScheduledTimeout = () => {
+    if (isScheduledAppLockEnabled()) {
+      window.clearTimeout(this.scheduledTimeOutId);
+      this.scheduledTimeOutId = window.setTimeout(this.showAppLock, getScheduledAppLockTimeOut() * 1000);
```

Comment:
Why `* 1000`? Can we add something to the naming scheme to identify the time unit?